Declarative Web Push: Support `pushnotification` event for declarative messages that are mutable

beidson opened the pull request:
#### 4b36aea360ae3875d94060a3c6705a7b3857b847
<pre>
Declarative Web Push: Support `pushnotification` event for declarative messages that are mutable
<a href="https://bugs.webkit.org/show_bug.cgi?id=261956">https://bugs.webkit.org/show_bug.cgi?id=261956</a>
rdar://114886688

Reviewed by Brent Fulgham and Chris Dumez.

When a declarative web push message arrives, and the sender has explicitly marked it &quot;mutable&quot;,
we fire a new `pushnotification` event at the Service Worker if there is one.

That event handler is given the proposed Notification described by the declarative message, and
has the opportunity to offer a new one instead.

It is also given the proposed AppBadge count described by the declarative message, and can modify
that as well.

If the event takes too long, or does *not* override the proposed Notification or AppBadge count,
then the originals described by the declarative message are used instead.

Implementation-wise, it&apos;s amazing how long of a chain of message handlers and completion handlers
exist to get a push message from the UI process to the point where an event can be dispatched
to the ServiceWorker.

This patch changes every single link in that chain to:
1 - Pass along the proposed NotificationPayload if one exists
2 - Change the CompletionHandlers to return the new NotificationPayload to be used.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::create):
(WebCore::Notification::Notification):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/NotificationOptionsPayload.h:
(WebCore::NotificationOptionsPayload::isolatedCopy):
* Source/WebCore/Modules/notifications/NotificationPayload.cpp:
(WebCore::NotificationPayload::fromNotificationData):
* Source/WebCore/Modules/notifications/NotificationPayload.h:
(WebCore::NotificationPayload::isolatedCopy):
* Source/WebCore/Modules/push-api/PushNotificationEvent.cpp: Copied from Source/WebCore/Modules/notifications/NotificationPayload.cpp.
(WebCore::PushNotificationEvent::PushNotificationEvent):
(WebCore::PushNotificationEvent::~PushNotificationEvent):
(WebCore::PushNotificationEvent::create):
* Source/WebCore/Modules/push-api/PushNotificationEvent.h: Added.
* Source/WebCore/Modules/push-api/PushNotificationEvent.idl: Copied from Source/WebCore/Modules/notifications/NotificationOptionsPayload.h.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.h: Copied from Source/WebCore/Modules/notifications/NotificationOptionsPayload.h.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.idl: Copied from Source/WebCore/Modules/notifications/NotificationOptionsPayload.h.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
(WebCore::ServiceWorkerInternals::schedulePushEvent):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::dispatchPushEvent):
(WebCore::ServiceWorkerGlobalScope::dispatchPushNotificationEvent):
(WebCore::ServiceWorkerGlobalScope::clearPushNotificationEvent):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/context/SWContextManager.cpp:
(WebCore::SWContextManager::firePushEvent):
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFirePushEvent):
(WebCore::ServiceWorkerThread::queueTaskToFirePushNotificationEvent):
(WebCore::ServiceWorkerThread::startNotificationPayloadFunctionalEventMonitoring):
(WebCore::ServiceWorkerThread::heartBeatTimerFired):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
(WebCore::ServiceWorkerThread::stopNotificationPayloadFunctionalEventMonitoring):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::firePushEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::processPushMessage):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::processPushMessage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::firePushEvent):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Cocoa/WebPushMessageCocoa.mm:
(WebKit::WebPushMessage::fromDictionary):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPushMessage.h:
* Source/WebKit/Shared/WebPushMessage.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::processPushMessage):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::firePushEvent):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::didReceivePushMessage):
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::getPendingPushMessages):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(-[PushNotificationDelegate clearMostRecents]):

Canonical link: <a href="https://commits.webkit.org/268358@main">https://commits.webkit.org/268358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de03b84e804a1f9431228638fdafcdbf9fbc8844

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19437 "24 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21326 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19674 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23119 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19998 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19783 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16889 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22183 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19616 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16867 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17677 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23988 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17852 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21959 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18453 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15622 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17597 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4651 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/notifications/Notification.h:
```diff
@@ -36,6 +36,7 @@
 #include "ActiveDOMObject.h"
 #include "EventTarget.h"
 #include "NotificationDirection.h"
+#include "NotificationPayload.h"
```
This could just be forward declared, since it's passed as a reference and not a member of this class.

beidson replied:
That is how it originally was. In one of the EWS runs it broke various different builds in fun and creative ways.
Because it was building locally, and shrouded in UnifiedSource mystery on the bots... I decided to take the shortcut option. =/

brentfulgham commented on Source/WebCore/Modules/push-api/PushNotificationEvent.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+}
+
+PushNotificationEvent::~PushNotificationEvent()
```
Could this be `~PushNotificationEvent() = default;`?

beidson replied:
Yup. Needed to set a breakpoint at some point, didn't revert back 👍

cdumez replied:
I think you missed this change.

brentfulgham commented on Source/WebCore/Modules/push-api/PushNotificationEvent.h:
```diff
@@ -0,0 +1,67 @@
+#include "ExtendableEvent.h"
+#include "Notification.h"
+#include "NotificationData.h"
+#include "PushNotificationEventInit.h"
```
I think these last three includes could be forward-declared.

brentfulgham commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -311,7 +311,21 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
+    // If we're handling a PushNotificationEvent, this Notification will override the proposed notification
+    // instead of being shown directly.
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    if (serviceWorkerGlobalScope) {
```
You could probably move the `#if ENABLE(DECLARATIVE_WEB_PUSH)` inside the `if(serviceWorkerGlobalScope)`, then you could delete line 328.

brentfulgham commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2503,34 +2503,40 @@ void NetworkProcess::processPushMessage(PAL::SessionID sessionID, WebPushMessage
+        bool isDeclarative = !!pushMessage.notificationPayload;
+        session->ensureSWServer().processPushMessage(WTFMove(pushMessage.pushData), WTFMove(pushMessage.notificationPayload), WTFMove(pushMessage.registrationURL), [this, protectedThis = Ref { *this }, sessionID, origin = WTFMove(origin), scope = WTFMove(scope), callback = WTFMove(callback), isDeclarative](bool result, std::optional<WebCore::NotificationPayload>&& resultPayload) mutable {
+
+            if (!isDeclarative) {
```
Maybe:
```
if (!isDeclarative && !result) {
    if (auto session = networkSession(sessionID)) {
        session->notificationmanager()....
```

brentfulgham commented on Source/WebKit/Shared/WebPushMessage.h:
```diff
@@ -25,13 +25,11 @@
 
 #pragma once
 
+#include <WebCore/NotificationPayload.h>
```
Could `WebCore::NotificationPayload` be forward declared?

beidson replied:
Of all the "This could be forward declared?" comments this one hurts the most in the number of files I'd have to go touch.

The frustration here is that I can get it building locally, but then the wildly different configs on the bots (iOS, tv, older Mac, linuxes) all have different notions of unified sources.

I tried to resolve these forward declare vs includes multiple times today, but can't afford to spend any more time back-and-forthing with EWS on this =(

brentfulgham replied:
If it's a huge issue, don't bother. I've just trained myself to watch for these includes since it's a lot of work to remove them later (as you know!) :-)

brentfulgham reviewed (approved):
Looks good, I had a few small questions and suggestions but no need for another review from me. r=me!

cdumez commented on Source/WebCore/Modules/push-api/PushNotificationEvent.h:
```diff
@@ -0,0 +1,67 @@
+    std::optional<uint64_t> proposedAppBadge() const { return m_proposedAppBadge; }
+
+    void setUpdatedNotificationData(NotificationData&& updatedData) { m_updatedNotificationData = WTFMove(updatedData); }
+    const std::optional<NotificationData> updatedNotificationData() const { return m_updatedNotificationData; }
```
Did you mean to return a `const std::optional<NotificationData>&` ?

beidson replied:
Yup, that's what I meant.

cdumez commented on Source/WebCore/Modules/push-api/PushNotificationEvent.h:
```diff
@@ -0,0 +1,67 @@
+    void setUpdatedAppBadge(std::optional<uint64_t>&& updatedAppBadge) { m_updatedAppBadge = WTFMove(updatedAppBadge); }
+    const std::optional<std::optional<uint64_t>> updatedAppBadge() const { return m_updatedAppBadge; }
+
+    EventInterface eventInterface() const final { return PushNotificationEventInterfaceType; }
```
Can probably be private.

beidson replied:
Yup.

cdumez commented on Source/WebCore/Modules/push-api/PushNotificationEvent.h:
```diff
@@ -0,0 +1,67 @@
+    RefPtr<Notification> m_proposedNotification;
+    std::optional<uint64_t> m_proposedAppBadge;
+    std::optional<NotificationData> m_updatedNotificationData;
+    std::optional<std::optional<uint64_t>> m_updatedAppBadge;
```
This double nesting is a little confusing.

cdumez replied:
I looked again and I don't see any good reason why we need 2 levels of optional here.

beidson replied:
Agreed, but it is necessary - An AppBadge is properly a `std::optional<uint64_t>` because "missing value" has an explicit meaning...
And here we need to track "Is there actually an updated one?"

i.e. `setUpdatedAppBadge(std::nullopt)` and "nobody set an updated AppBadge at all" both have distinct meanings that are captured by the double optional.

Another way of considering it:

`typedef std::optional<uint64_t> AppBadge`
`std::optional<AppBadge> m_updatedAppBadge`

cdumez commented on Source/WebCore/Modules/push-api/PushNotificationEventInit.h:
```diff
@@ -0,0 +1,41 @@
+
+struct PushNotificationEventInit : ExtendableEventInit {
+    RefPtr<Notification> proposedNotification;
+    unsigned long long proposedAppBadge;
```
Should be a `std::optional<unsigned long long>` since the member is optional.

beidson replied:
Yup

cdumez commented on Source/WebCore/bindings/js/JSPushNotificationEventCustom.cpp:
```diff
@@ -0,0 +1,52 @@
+JSC::JSValue JSPushNotificationEvent::proposedAppBadge(JSC::JSGlobalObject&) const
+{
+    // FIXME: This custom getter implements the following IDL intention:
+    // readonly attribute unsigned long long? proposedAppBadge;
```
Are you sure this isn't supported? I thought this worked, as long as the implementation returned a `std::optional<unsigned long>`.

cdumez replied:
I just checked and this is totally supported. We have plenty of examples in the existing code base where this is working. We don't need custom bindings.

beidson replied:
When I figured tried, it absolutely failed compiling miserably.

I *do* see other users of it, so I'm trying again.

cdumez replied:
Ping me on Slack if you get into trouble with the generated bindings. Hopefully I can help. We should definitely be able to make this work without custom bindings.

cdumez reviewed (commented):
Still reviewing but putting a few comments out there since this was already r+

cdumez commented on Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
```diff
@@ -292,22 +292,22 @@ void WebSWContextManagerConnection::fireActivateEvent(ServiceWorkerIdentifier id
-        queue->dispatch([result, callback = WTFMove(callback)]() mutable {
-            callback(result);
+    auto inQueueCallback = [queue = m_queue, callback = WTFMove(callback)](bool result, std::optional<NotificationPayload>&& resultPayload) mutable {
+        queue->dispatch([result, resultPayload = WTFMove(resultPayload), callback = WTFMove(callback)]() mutable {
```
`WTFMove(resultPayload).isolatedCopy()` no? since we're passing to another thread.

brentfulgham replied:
I don't understand why we have code that does `WTFMove(resultPayload).isolatedCopy()`, if moving the object to another thread is bad.

Shouldn't such cases always just be `resultPayload = resultPayload.isolatedCopy()` (assuming the return of `isolatedCopy` is automatically moved to the new lambda value?

cdumez replied:
Many of our objects (like String) have a more optimized isolatedCopy() version when called on a r-value reference (which avoids isolated copying in cases where it is safe). Therefore, if our can WTFMove(), you should WTFMove().isolatedCopy().

Also, you should always isolatedCopy() when passing non-trivial types to another thread.

beidson replied:
Yah, if we deem somewhere needs an isolatedCopy(), then the WTFMove(...) part is unnecessary.

I'll check more closely on this, but I'm not convinced we need them:
It came directly off IPC and been WTFMove()'d around.

After addressing all other feedback, will look closely.

cdumez replied:
> Yah, if we deem somewhere needs an isolatedCopy(), then the WTFMove(...) part is unnecessary.

I explained that this isn't true. `WTFMove(foo).isolatedCopy()` is more efficient with MANY of our types. See for example String:
```
    WTF_EXPORT_PRIVATE String WARN_UNUSED_RETURN isolatedCopy() const &;
    WTF_EXPORT_PRIVATE String WARN_UNUSED_RETURN isolatedCopy() &&;
```

> It came directly off IPC and been WTFMove()'d around.

We use WTFMove().isolatedCopy() these days to do this safely. It is as efficient and you can't get it wrong. It is bad practice to make assumption and just WTFMove() to another thread. We stopped doing this a while back.

cdumez commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -360,28 +360,33 @@ void ServiceWorkerThreadProxy::didSaveScriptsToDisk(ScriptBuffer&& script, HashM
+    if (proposedPayload)
+        payloadCopy = *proposedPayload;
+
+    bool isPosted = postTaskForModeToWorkerOrWorkletGlobalScope([this, protectedThis = Ref { *this }, identifier, data = WTFMove(data), proposedPayload = WTFMove(proposedPayload)](auto&) mutable {
```
Don't we need to isolatedCopy() data and proposedPayload?

cdumez replied:
I looked at those types. We don't need to isolatedCopy() `data` but I believe we need it for `proposedPayload`.

beidson replied:
I don't think so - Each of them has directly come in off IPC and been WTFMove()'d around.

Certainly data never has been before this patch.

cdumez replied:
That's what WTFMove(proposedPayload).isolatedCopy() is for and you need to make sure proposedPayload has a && isolatedCopy().

cdumez commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -360,28 +360,33 @@ void ServiceWorkerThreadProxy::didSaveScriptsToDisk(ScriptBuffer&& script, HashM
+
+    bool isPosted = postTaskForModeToWorkerOrWorkletGlobalScope([this, protectedThis = Ref { *this }, identifier, data = WTFMove(data), proposedPayload = WTFMove(proposedPayload)](auto&) mutable {
+        thread().queueTaskToFirePushEvent(WTFMove(data), WTFMove(proposedPayload), [this, protectedThis = WTFMove(protectedThis), identifier](bool result, std::optional<NotificationPayload> resultPayload) mutable {
+            callOnMainThread([this, protectedThis = WTFMove(protectedThis), identifier, result, resultPayload = WTFMove(resultPayload)]() mutable {
```
Don't we need to isolatedCopy() resultPayload ?

beidson replied:
See above discussion

beidson replied:
(It's possible at some link in the chain I've introduced a new value to it, I'll be looking closely)

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -311,7 +311,21 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
+    // instead of being shown directly.
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    if (serviceWorkerGlobalScope) {
+        if (auto* pushNotificationEvent = serviceWorkerGlobalScope->pushNotificationEvent()) {
```
RefPtr

beidson replied:
Not needed? It's not stored for later use and cannot possibly be cleared here before this use is done.

cdumez replied:
Ryosuke and I are fixing them after you land code like this because the static analyzer complains. You're calling a function on it on the next line so you need to hold a strong ref to it. Recent policy change for safety.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -311,7 +311,21 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
     }
 
-    if (auto* serviceWorkerGlobalScope = dynamicDowncast<ServiceWorkerGlobalScope>(context)) {
+    auto* serviceWorkerGlobalScope = dynamicDowncast<ServiceWorkerGlobalScope>(context);
```
RefPtr

beidson replied:
Definitely less sure about lifecycle for this one - Added.

cdumez commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -66,6 +68,15 @@ void WorkerNavigator::setAppBadge(std::optional<unsigned long long> badge, Ref<D
 
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    if (is<ServiceWorkerGlobalScope>(*scope)) {
+        if (auto* pushNotificationEvent = downcast<ServiceWorkerGlobalScope>(*scope).pushNotificationEvent()) {
```
RefPtr

beidson replied:
For the same reason as the last one about the current notification even, not needed.

cdumez replied:
Please, no raw pointer variable in new code, even if you think it is safe.

cdumez commented on Source/WebCore/Modules/push-api/PushNotificationEvent.idl:
```diff
@@ -0,0 +1,38 @@
+] interface PushNotificationEvent {
+    constructor([AtomString] DOMString type, optional PushNotificationEventInit eventInitDict);
+    readonly attribute Notification proposedNotification;
+    [CustomGetter] readonly attribute unsigned long long proposedAppBadge;
```
Missing `?` since you say in the comment below that this attribute is nullable.

cdumez commented on Source/WebCore/Modules/push-api/PushNotificationEventInit.idl:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2023?

cdumez commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -66,6 +68,15 @@ void WorkerNavigator::setAppBadge(std::optional<unsigned long long> badge, Ref<D
 
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    if (is<ServiceWorkerGlobalScope>(*scope)) {
+        if (auto pushNotificationEvent = RefPtr { downcast<ServiceWorkerGlobalScope>(*scope).pushNotificationEvent() }) {
```
nit: little bit of a weird pattern. This would be more concise and achieve the same:
```
if (RefPtr event = downcast<ServiceWorkerGlobalScope>(*scope).pushNotificationEvent()) {
```

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -311,7 +311,19 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
     }
 
-    if (auto* serviceWorkerGlobalScope = dynamicDowncast<ServiceWorkerGlobalScope>(context)) {
+    auto serviceWorkerGlobalScope = RefPtr { dynamicDowncast<ServiceWorkerGlobalScope>(context) };
```
Same weird pattern. Not much use in saying both auto and RefPtr.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -311,7 +311,19 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
+    // instead of being shown directly.
+    if (serviceWorkerGlobalScope) {
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+        if (auto pushNotificationEvent = RefPtr { serviceWorkerGlobalScope->pushNotificationEvent() }) {
```
ditto

cdumez commented on Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
```diff
@@ -254,11 +266,58 @@ void ServiceWorkerThread::queueTaskToFirePushEvent(std::optional<Vector<uint8_t>
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+void ServiceWorkerThread::queueTaskToFirePushNotificationEvent(NotificationPayload&& proposedPayload, Function<void(bool, std::optional<NotificationPayload>&&)>&& callback)
+{
+    auto& serviceWorkerGlobalScope = downcast<ServiceWorkerGlobalScope>(*globalScope());
```
Ref

cdumez commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -360,28 +360,33 @@ void ServiceWorkerThreadProxy::didSaveScriptsToDisk(ScriptBuffer&& script, HashM
+
+    bool isPosted = postTaskForModeToWorkerOrWorkletGlobalScope([this, protectedThis = Ref { *this }, identifier, data = crossThreadCopy(WTFMove(data)), proposedPayload = crossThreadCopy(WTFMove(proposedPayload))](auto&) mutable {
+        thread().queueTaskToFirePushEvent(WTFMove(data), WTFMove(proposedPayload), [this, protectedThis = WTFMove(protectedThis), identifier](bool result, std::optional<NotificationPayload> resultPayload) mutable {
+            callOnMainThread([this, protectedThis = WTFMove(protectedThis), identifier, result, resultPayload = WTFMove(resultPayload)]() mutable {
```
Missing crossThreadCopy() for resultPayload.

cdumez reviewed (approved):
LGTM with nits.

Nice that you were able to make `unsigned long long?` work in the IDL.

beidson committed:
Declarative Web Push: Support `pushnotification` event for declarative messages that are mutable
https://bugs.webkit.org/show_bug.cgi?id=261956
rdar://114886688

Reviewed by Brent Fulgham and Chris Dumez.

When a declarative web push message arrives, and the sender has explicitly marked it "mutable",
we fire a new `pushnotification` event at the Service Worker if there is one.

That event handler is given the proposed Notification described by the declarative message, and
has the opportunity to offer a new one instead.

It is also given the proposed AppBadge count described by the declarative message, and can modify
that as well.

If the event takes too long, or does *not* override the proposed Notification or AppBadge count,
then the originals described by the declarative message are used instead.

Implementation-wise, it's amazing how long of a chain of message handlers and completion handlers
exist to get a push message from the UI process to the point where an event can be dispatched
to the ServiceWorker.

This patch changes every single link in that chain to:
1 - Pass along the proposed NotificationPayload if one exists
2 - Change the CompletionHandlers to return the new NotificationPayload to be used.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::create):
(WebCore::Notification::Notification):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/NotificationOptionsPayload.h:
(WebCore::NotificationOptionsPayload::isolatedCopy):
* Source/WebCore/Modules/notifications/NotificationPayload.cpp:
(WebCore::NotificationPayload::fromNotificationData):
* Source/WebCore/Modules/notifications/NotificationPayload.h:
(WebCore::NotificationPayload::isolatedCopy):
* Source/WebCore/Modules/push-api/PushNotificationEvent.cpp: Copied from Source/WebCore/Modules/notifications/NotificationPayload.cpp.
(WebCore::PushNotificationEvent::PushNotificationEvent):
(WebCore::PushNotificationEvent::~PushNotificationEvent):
(WebCore::PushNotificationEvent::create):
* Source/WebCore/Modules/push-api/PushNotificationEvent.h: Added.
* Source/WebCore/Modules/push-api/PushNotificationEvent.idl: Copied from Source/WebCore/Modules/notifications/NotificationOptionsPayload.h.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.h: Copied from Source/WebCore/Modules/notifications/NotificationOptionsPayload.h.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.idl: Copied from Source/WebCore/Modules/notifications/NotificationOptionsPayload.h.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
(WebCore::ServiceWorkerInternals::schedulePushEvent):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::dispatchPushEvent):
(WebCore::ServiceWorkerGlobalScope::dispatchPushNotificationEvent):
(WebCore::ServiceWorkerGlobalScope::clearPushNotificationEvent):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/context/SWContextManager.cpp:
(WebCore::SWContextManager::firePushEvent):
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFirePushEvent):
(WebCore::ServiceWorkerThread::queueTaskToFirePushNotificationEvent):
(WebCore::ServiceWorkerThread::startNotificationPayloadFunctionalEventMonitoring):
(WebCore::ServiceWorkerThread::heartBeatTimerFired):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
(WebCore::ServiceWorkerThread::stopNotificationPayloadFunctionalEventMonitoring):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::firePushEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::processPushMessage):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::processPushMessage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::firePushEvent):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Cocoa/WebPushMessageCocoa.mm:
(WebKit::WebPushMessage::fromDictionary):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPushMessage.h:
* Source/WebKit/Shared/WebPushMessage.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::processPushMessage):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::firePushEvent):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::didReceivePushMessage):
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::getPendingPushMessages):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(-[PushNotificationDelegate clearMostRecents]):

Canonical link: https://commits.webkit.org/268358@main
